torch/fx/experimental/gradual_graph_typechecker.py

```diff
@@ -0,0 +1,90 @@
+from functools import reduce
+from torch.fx.tensor_type import Dyn, consistency, Tensor_Type
+
+
+def type_check(graph):
+    """
+    A gradual type checker for graphs
+    """
+    env = {}
+    expr_env = {}
+    # populate the type environment
+    collect_params(graph.nodes, env)
+    for n in graph.nodes:
+        if not type_check_node(n, env, expr_env):
+            return False
+    return True
+
+
+def collect_params(node_list, env):
+    """
+    Collect argument types from the type environment.
+    If not present, populate with Dyn
+    """
+    for node in node_list:
+        if node.op == 'placeholder':
+            if node.type:
+                env[node.name] = node.type
+            else:
+                node.type = Dyn
+                env[node.name] = node.type
+
+
+def type_check_node(n, env, expr_env):
+    """
+    Type check a given node.
+    Current operations:
+    - Reshape
+    - Transpose
+    - Add
+    """
+    if n.op == 'placeholder':
+        if n.name in env.keys():
+            n.type = env[n.name]
+            return n.type
+        else:
+            raise Exception("free variable")
+
+    if n.op == 'call_function':
+        if n.name == 'add':
+            t1 = type_check_node(n.args[0], env, expr_env)
```

Comment:
Can you explain the control flow here? We're already applying `type_check_node` on every node in `type_check`, but then we're recursively calling it as well on all of the argument nodes for each of these rules. At the very least this is redundant iiuc, unless this is side-effectful?

Is it sufficient to just type check the nodes in the topological order of the program? or do we need this recursion

Reply:
I thought more about this. The arguments are variables. We can either typecheck them by looking them up in the environment or we can typecheck them by calling the typechecker again which will go through the variable case and typecheck them. I can change it for now to just lookup in the type environment